sno/dataset2.py

```diff
@@ -0,0 +1,218 @@
+            [schema-x-hash]     = [schema-x JSON]
+            [schema-y-hash]     = [schema-y JSON]
+            ...
+
+        [hex(pk-hash):2]/
+          [hex(pk-hash):2]/
+            [base64(pk-value)]  = [msgpack([schema-z-hash, value0, value1, ...])]
+
+    Dataset2 is initialised pointing at a particular directory tree, and uses that
+    to read features and schemas. However, it never writes to the tree, since this
+    is not straight-forward in git/sno and involves batching writes into a commit.
+    Therefore, there are no methods which write, only methods which return things
+    which *should be written*. The caller must write these to a commit.
+    """
+
+    def __init__(self, tree):
+        # Generally a pygit2.Tree, but conceptually just a directory tree
+        # with the appropriate structure.
+        self.tree = tree
+        # TODO - add support for reading from pygit2.Index or similar during merge.
+
+    def get_data_at(self, path):
+        """Return the data at the given path from within this dataset."""
+        leaf = self.tree / str(path)
+        if not hasattr(leaf, "data"):
+            raise IntegrityError(f"No data found at path {path}, type={type(leaf)}")
+        return leaf.data
+
+    @functools.lru_cache(maxsize=16)
+    def get_schema(self, schema_hash):
+        """Load the schema with the given hash from this dataset."""
+        return Schema.loads(self.get_data_at(f"meta/schema/{schema_hash}"))
+
+    @classmethod
+    def encode_schema(cls, schema):
+        """
+        Given a schema, returns the path and the data which *should be written*
+        to write this schema. This is almost the inverse of get_schema, except
+        Dataset2 doesn't write the data.
+        """
+        return f"meta/schema/{schema.principal_hash()}", schema.dumps()
+
+    def read_feature_to_tuple(self, path):
+        """
+        Given a dataset path, reads the feature stored there,
+        and returns the feature row as a tuple.
+        """
+        data = self.get_data_at(path)
+        schema_hash, non_pk_values = _unpack(data)
+        schema = self.get_schema(schema_hash)
+        pk_values = self.decode_path_to_pk_values(path)
+        feature = schema.merge_pk_and_non_pk(pk_values, non_pk_values)
+        return feature
+
+    @classmethod
+    def decode_path_to_pk_values(cls, path):
+        """Given a feature path, returns the pk values encoded in it."""
+        encoded = os.path.basename(path)
+        return _unpack(_b64decode_str(encoded))
+
+    @classmethod
+    def encode_feature_tuple(cls, feature, schema):
+        """
+        Given a feature row and a schema, returns the path and the data which
+        *should be written* to write this feature. This is almost the inverse
+        of read_feature_to_tuple, except Dataset2 doesn't write the data.
+        """
+        pk_values, non_pk_values = schema.split_pk_and_non_pk(feature)
+        path = cls.encode_pk_values_to_path(pk_values)
```

Comment:
Need to think through what this means for datasets without primary keys.

In principle I think: 
1. you can only bulk-swap the entire dataset in a commit. The blob store means most updates don't change anything, but there can't be a simple concept of insert/update/delete on a row level. #112 
2. essentially our key has to be the same as git uses for the blob — the content of the feature/row.
3. if we want it to fit into the existing feature tree structure I guess we need a way of telling actual PKs apart from a hash of the row-content? eg: rather than encoding the pk as `_pack([123])` we use a custom msgpack type (for hash bytes) similar to how we handle geometries in Dataset1.

Thoughts @craigds? This needs to be fleshed out in the spec and is a larger thing to consider.

AFAICT at the moment this code will encode all features for a dataset without PKs using `[]`, which presumably will mean there's only one row stored. Maybe change it to explicitly assert for the meantime?